write-fonts/src/graph.rs

```diff
@@ -809,6 +820,210 @@ impl Graph {
+                subtable_ref.object = ext_id;
+            }
+            lookup.write_over(lookup_type.promote().to_raw(), 0);
+            lookup.type_ = lookup_type.promote().into();
+            lookup.name = "ExtensionLookup";
+            self.objects.insert(*id, lookup);
+        }
+        self.parents_invalid = true;
+        self.positions_invalid = true;
+    }
+
+    // get the list of tables that can be promoted, as well as the id of their parent table
+    fn get_promotable_subtables(&self) -> Option<(Vec<ObjectId>, ObjectId)> {
+        let can_promote = self
+            .objects
+            .iter()
+            .filter_map(|(id, obj)| (obj.type_.is_promotable()).then_some(*id))
+            .collect::<Vec<_>>();
+
+        if can_promote.is_empty() {
+            return None;
+        }
+
+        // sanity check: ensure that all promotable tables have a common root.
+        let parents = can_promote
+            .iter()
+            .flat_map(|id| {
+                self.nodes
+                    .get(id)
+                    .expect("all nodes exist")
+                    .parents
+                    .iter()
+                    .map(|x| x.0)
+            })
+            .collect::<HashSet<_>>();
+
+        // the only promotable subtables should be lookups, and there should
+        // be a single LookupList that is their parent; if there is more than
+        // one parent then something weird is going on.
+        if parents.len() > 1 {
+            if cfg!(debug_assertions) {
+                panic!("Promotable subtables exist with multiple parents");
+            } else {
+                log::warn!("Promotable subtables exist with multiple parents");
+                return None;
+            }
+        }
+
+        let parent_id = *parents.iter().next().unwrap();
+        Some((can_promote, parent_id))
+    }
+
+    /// select the tables to promote to extension, harfbuzz algorithm
+    ///
+    /// Based on the logic in HarfBuzz's [`_promote_exetnsions_if_needed`][hb-promote][hb-promote] function.
+    ///
+    /// [hb-promote]: https://github.com/harfbuzz/harfbuzz/blob/5d543d64222c6ce45332d0c188790f90691ef112/src/hb-repacker.hh#L97
+    fn select_promotions_hb(&self, candidates: &[ObjectId], parent_id: ObjectId) -> Vec<ObjectId> {
+        struct LookupSize {
+            id: ObjectId,
+            subgraph_size: usize,
+            subtable_count: usize,
+        }
+
+        impl LookupSize {
+            // I could impl Ord but then I need to impl PartialEq and it ends
+            // up being way more code
+            fn sort_key(&self) -> impl Ord {
+                let bytes_per_subtable = self.subtable_count as f64 / self.subgraph_size as f64;
+                // f64 isn't ord, so we turn it into an integer,
+                // then reverse, because we want bigger things first
+                std::cmp::Reverse((bytes_per_subtable * 1e9) as u64)
+            }
```

Comment:
Seems like HB falls back to comparing lookup list order when the bytes per subtable sizes are equal. I'm not sure how common this is and it doesn't affect correctness, but may be something to keep in mind if we're directly comparing output and find discrepancies.

Reply:
Good point. cc @garretrieger, is there any particular rationale for explicitly using lookup index as a tie-break? I mean here:

https://github.com/harfbuzz/harfbuzz/blob/4056315c868693c655373856a83603fdc1587aab/src/hb-repacker.hh#L56-L69

Reply:
I don't think this isn't particularly important from a repacking perspective, but using lookup idx to break ties should guarantee a stable sort order. ie. you're not at the mercy of the chosen sorting algorithm to order tie breaks for you.